File: Week 3 Assignment - Kavya.py
import numpy as np
import pandas as pd
import time

# Read the Excel file
file_path = "clinics.xls"
df = pd.read_excel(file_path)

# Define Haversine distance function
def haversine(lat1, lon1, lat2, lon2):
    MILES = 3959  # Earth radius in miles
    lat1, lon1, lat2, lon2 = map(np.radians, [lat1, lon1, lat2, lon2])
    dlat = lat2 - lat1
    dlon = lon2 - lon1
    a = np.sin(dlat/2)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon/2)**2
    return MILES * (2 * np.arcsin(np.sqrt(a)))

# Define a reference location (example: central NYC)
REF_LAT, REF_LONG = 40.7128, -74.0060  # Adjust based on need

# Method 1: Using a for loop
def haversine_loop(df):
    distances = []
    for _, row in df.iterrows():
        distances.append(haversine(REF_LAT, REF_LONG, row['locLat'], row['locLong']))
    return distances

# Method 2: Using apply function
def haversine_apply(df):
    return df.apply(lambda row: haversine(REF_LAT, REF_LONG, row['locLat'], row['locLong']), axis=1)

# Method 3: Using NumPy vectorization
def haversine_vectorized(df):
    return haversine(REF_LAT, REF_LONG, df['locLat'].values, df['locLong'].values)

# Measure execution times
start = time.time()
df['distance_loop'] = haversine_loop(df)
print("For-loop time:", time.time() - start)

start = time.time()
df['distance_apply'] = haversine_apply(df)
print("Apply function time:", time.time() - start)

start = time.time()
df['distance_vectorized'] = haversine_vectorized(df)
print("Vectorized NumPy time:", time.time() - start)

File: Week 3 Assignment - Kavya.R
# Load necessary libraries
library(geosphere)
library(dplyr)

# Read the Excel file
library(readxl)
file_path <- "clinics.xls"
df <- read_excel(file_path)

df <- df %>%
  mutate(
    locLat = as.numeric(locLat),
    locLong = as.numeric(locLong)
  )

# Define the Haversine function
haversine <- function(lat1, lon1, lat2, lon2) {
  R <- 3959
  lat1 <- lat1 * pi / 180
  lon1 <- lon1 * pi / 180
  lat2 <- lat2 * pi / 180
  lon2 <- lon2 * pi / 180
  
  dlat <- lat2 - lat1
  dlon <- lon2 - lon1
  a <- sin(dlat / 2)^2 + cos(lat1) * cos(lat2) * sin(dlon / 2)^2
  c <- 2 * asin(sqrt(a))
  return(R * c)
}

# Reference location
ref_lat <- 40.7128
ref_lon <- -74.0060

# Method 1: For-loop approach (Fixed)
start_time <- Sys.time()
df$distance_loop <- sapply(1:nrow(df), function(i) {
  if (is.na(df$locLat[i]) || is.na(df$locLong[i])) {
    return(NA)
  }
  haversine(ref_lat, ref_lon, df$locLat[i], df$locLong[i])
})
end_time <- Sys.time()
cat("For-loop time:", end_time - start_time, "seconds\n")

# Display first few rows to verify
head(df)

# Method 1: For-loop approach
start_time <- Sys.time()
df$distance_loop <- sapply(1:nrow(df), function(i) {
  haversine(ref_lat, ref_lon, df$locLat[i], df$locLong[i])
})
end_time <- Sys.time()
cat("For-loop time:", end_time - start_time, "seconds\n")

# Method 2: Using apply()
start_time <- Sys.time()
df$distance_apply <- apply(df, 1, function(row) {
  haversine(ref_lat, ref_lon, as.numeric(row["locLat"]), as.numeric(row["locLong"]))
})
end_time <- Sys.time()
cat("Apply function time:", end_time - start_time, "seconds\n")

# Method 3: Vectorized approach using geosphere package
start_time <- Sys.time()
df$distance_vectorized <- distHaversine(
  matrix(c(df$locLong, df$locLat), ncol = 2),
  matrix(c(ref_lon, ref_lat), ncol = 2)
) / 1609.34
end_time <- Sys.time()
cat("Vectorized approach time:", end_time - start_time, "seconds\n")
